### ISSUE TITLE
Support owner, group, permissions, and timestamps on transfers

### ISSUE THREAD

**Issue by adammoody** [LLNL/UnifyFS#435]
While we don't necessarily enforce permission bits for files within UnifyFS, users will want to keep any permissions they have set when files are transferred to the parallel file system.  To do this we need to:
1) Add wrappers for chown and utimes, perhaps others
2) Update our transfer utility to read these fields and set them on the resulting file on the parallel file system

**Comment by adammoody** [comment]
@craigsteffen , since you have been working on the transfer tool, are you game for making these changes?  I think item 2 should be easy.  The first item is bit more work, mainly because there is a learning curve involved to add new wrappers.

**Comment by adammoody** [comment]
On the transfer utility, we'd want to call stat() to get these values, then apply them to the file on the parallel file system with calls like chmod(), chown(), and utimes().

**Comment by craigsteffen** [comment]
(from LLNL call on 2020 jan 22)
this is probably at least three items, and we can split this out into separate issues.  
1) when files are copied from external file system into the UnifyFS (using the transfer API), get, preserve, and store those permissions (and timestamps and owner and so on) when that file is instantiated within unifyFS.  (copying files in will probably be a rare case but we should support it)
2) when a user opens a file within unify, they may use calls like chmod or utime or the like to manipulate the file.  We need to support that. 
3) when a file is transferred from within the unify fs out to the external file system (using the transfer API), the permission bits and timestamps and owner and so on should be applied to the new external file.

**Comment by craigsteffen** [comment]
> On the transfer utility, we'd want to call stat() to get these values, then apply them to the file on the parallel file system with calls like chmod(), chown(), and utimes().

Right, but we'll want to be very careful to also handle function calls like chown() that will frequently fail.  chowning the file to someone else, for instance.  Which actually brings up how should we treat that within unify?  Because that's typically not allowed in a unix file system; to chown a file to someone else.

**Comment by craigsteffen** [comment]
Anyway, I'll need to make a list or a matrix of what things we might support, and which ones will likely actually do something useful, and what the failure will actually do.

**Comment by craigsteffen** [comment]
I'll start working on this when I'm back from the ECP meeting.  I agree; item 2 should be fairly straightforward; run a stat() of the file and transfer the data in.  We'll have to make a priority list for the functions in item 1, so that we create them in a logical order.

**Comment by craigsteffen** [comment]
To-do items on this.  

- When copying files into Unifyfs space, we need to capture the permissions of the file and at least store them (if not honor them) while the file exists in UnifyFS space.  

- We need to decide what externally-facing metadata we we will store for files in UnifyFS, and a subset of that, which ones we will honor, and also, which things will we allow user to manipulate with standard tools on files within UnifyFS.  These should be documented.  

- When copying files out of UnifyFS, we need to capture permission bits (and other ownership information) from the file as it exists within UnifyFS and transfer those to the external file system.  Or not.  But we need to document what we do and do not transfer.  

I presume the answer for the moment is that the "write" bit is very definitely tracked, and can be manipulated (for lamination reasons), but the others will not be tracked or propagated.  

The path for this will first to create a place to store the file metadata flags for each file in the metadata store.  Then we'll create the functions to use and manipulate them.  

Before that happens, then we should document as follows: 

All Files will effectively have permissions:
rw-------
before lamination and 
r--------
after lamination. 
We assume that until specific steps are made, that files transferred out of UnifyFS to the parallel file system will have whatever permission bits the umask of the running process has, has the owner ID of the owner of that process, and the group of the effective group ID of that process.
